File: app/lib/engine/draw_brain.dart
import 'dart:math';

import 'model.dart';
import 'scoring.dart';

/// V2.2 §32–39 anti-starvation draw system.
///
/// The stock stays shuffled, but every stock draw examines a small
/// look-ahead window (default 9) and picks the most useful card for the
/// drawing player. Utility rewards: group-formers > pair-formers > score
/// reduction > low-value cards. Dry-draw soft pity gradually shifts the
/// bias toward useful cards after repeated unproductive draws; it never
/// guarantees a specific rank.
///
/// Opening balancer (§38–39): a fresh hand with no duplicate rank has a
/// configurable chance of being nudged toward at least one pair.
abstract final class DrawBrain {
  /// V2.2: examines the top N stock cards to smooth out dry streaks.
  static const int lookAheadWindow = 9;

  /// Dry draws below this count use pure utility ordering.
  static const int dryDrawThreshold = 3;

  /// Dry draws above this cap deliver a maximum-boost pick.
  static const int dryDrawCap = 6;

  /// A drawn card is "productive" when it beats this delta (score reduction).
  static const int productiveDelta = 1;

  /// V2.2 §38 opening balancer: chance to nudge a dead opening toward a pair.
  static const double openingBalancerChance = 0.72;

  // ---------- stock draw ------------------------------------------------

  /// Choose which card to serve from [stock] to [hand]. The chosen card is
  /// removed from [stock] (which uses last-element = top ordering) and the
  /// remainder keeps its relative order. [dryStreak] biases the pick toward
  /// higher-utility cards after repeated unproductive draws.
  ///
  /// Returns the served card. Does NOT modify [hand] — the caller adds it.
  static Card drawFromStock(List<Card> stock, Hand hand, int dryStreak) {
    if (stock.isEmpty) {
      throw StateError('DrawBrain.drawFromStock: empty stock');
    }
    if (stock.length == 1) return stock.removeLast();

    final windowSize = min(lookAheadWindow, stock.length);
    // Stock top = last element; window is the last [windowSize] entries.
    final start = stock.length - windowSize;
    var bestIdx = stock.length - 1;
    var bestScore = _score(stock[bestIdx], hand, dryStreak);
    for (var i = start; i < stock.length - 1; i++) {
      final s = _score(stock[i], hand, dryStreak);
      // Prefer higher score; break ties by staying closer to the top.
      if (s > bestScore) {
        bestScore = s;
        bestIdx = i;
      }
    }
    return stock.removeAt(bestIdx);
  }

  /// Returns true if drawing [drawn] into [hand] meaningfully lowered the
  /// player's optimised count. Used by the session to increment or reset the
  /// per-player dry-draw counter.
  static bool wasProductive(Card drawn, Hand handBeforeDraw) {
    final before = ScoringEngine.count(handBeforeDraw.cards);
    // "Best if we discarded the drawn card immediately" is the worst case.
    final withDraw = [...handBeforeDraw.cards, drawn];
    var bestAfter = 1 << 30;
    for (var i = 0; i < withDraw.length; i++) {
      final rest = [...withDraw]..removeAt(i);
      final c = ScoringEngine.count(rest);
      if (c < bestAfter) bestAfter = c;
    }
    return before - bestAfter >= productiveDelta;
  }

  // ---------- opening balancer -----------------------------------------

  /// V2.2 §38–39: if [hand] has no pair, with [openingBalancerChance] search
  /// the remaining [stock] for a card whose rank matches one already in the
  /// hand and swap it with the highest-value card. Never guarantees a pair
  /// (only nudges), preserves card conservation, and only mutates on hit.
  static void balanceOpeningHand(Hand hand, List<Card> stock, Random rng) {
    if (hand.size == 0 || stock.isEmpty) return;
    if (_hasAnyPair(hand)) return;
    if (rng.nextDouble() >= openingBalancerChance) return;

    // Ranks the player already holds — any duplicate makes an instant pair.
    final ranksHeld = <Rank>{for (final c in hand.cards) c.rank};
    // Find any matching-rank card in the stock; first match is fine.
    var stockIdx = -1;
    for (var i = stock.length - 1; i >= 0; i--) {
      if (!stock[i].isSpecial && ranksHeld.contains(stock[i].rank)) {
        stockIdx = i;
        break;
      }
    }
    if (stockIdx < 0) return;

    // Swap with the highest face-value non-special card in hand.
    Card? victim;
    for (final c in hand.cards) {
      if (c.isSpecial) continue;
      if (victim == null || c.value > victim.value) victim = c;
    }
    if (victim == null || victim.value < stock[stockIdx].value) return;

    final promoted = stock[stockIdx];
    hand.remove(victim);
    hand.add(promoted);
    stock[stockIdx] = victim;
  }

  // ---------- internals --------------------------------------------------

  /// Utility score for [c] entering [hand]. Higher = more useful.
  ///  - completes a 3+ group (rank already has ≥2)                 +40
  ///  - forms a fresh pair (rank has exactly 1)                    +25
  ///  - would complete a pair for a held Special                   +30
  ///  - low face value (A, 2, 3)                                    +6
  ///  - mid face value (4-6)                                        +2
  ///  - high value with no partner already                          −6
  ///  - dry-pity bonus: scales useful picks after repeated dry draws
  static int _score(Card c, Hand hand, int dryStreak) {
    final rankCount = <Rank, int>{};
    var hasSpecial = false;
    for (final h in hand.cards) {
      if (h.isSpecial) {
        hasSpecial = true;
        continue;
      }
      rankCount[h.rank] = (rankCount[h.rank] ?? 0) + 1;
    }

    var utility = 0;
    var useful = false;

    if (c.isSpecial) {
      // A Special is only useful when the hand has an exact pair to consume.
      final hasExactPair = rankCount.values.any((n) => n == 2);
      utility += hasExactPair ? 35 : -10;
      useful = hasExactPair;
    } else {
      final n = rankCount[c.rank] ?? 0;
      if (n >= 2) {
        utility += 40;
        useful = true;
      } else if (n == 1) {
        utility += 25;
        useful = true;
        if (hasSpecial) utility += 30; // completes the Special's pair target
      } else if (c.value <= 3) {
        utility += 6;
        useful = true;
      } else if (c.value <= 6) {
        utility += 2;
      } else {
        utility -= 6;
      }
    }

    if (dryStreak >= dryDrawThreshold && useful) {
      final level = min(dryStreak, dryDrawCap) - dryDrawThreshold + 1;
      utility += level * 8;
    }
    return utility;
  }

  static bool _hasAnyPair(Hand hand) {
    final counts = <Rank, int>{};
    for (final c in hand.cards) {
      if (c.isSpecial) continue;
      counts[c.rank] = (counts[c.rank] ?? 0) + 1;
      if (counts[c.rank]! >= 2) return true;
    }
    return false;
  }
}

File: backend/game-engine/src/main/java/com/zerocount/engine/session/DrawBrain.java
package com.zerocount.engine.session;

import com.zerocount.engine.model.Card;
import com.zerocount.engine.model.Hand;
import com.zerocount.engine.model.Rank;
import com.zerocount.engine.scoring.ScoringEngine;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

/**
 * V2.2 §32–39 anti-starvation draw brain — Java parity of Dart {@code
 * DrawBrain}. See {@code app/lib/engine/draw_brain.dart} for the design
 * doc. Weights, thresholds, and behaviour must stay in sync so live matches
 * feel identical to offline.
 */
public final class DrawBrain {

    public static final int LOOK_AHEAD_WINDOW = 9;
    public static final int DRY_DRAW_THRESHOLD = 3;
    public static final int DRY_DRAW_CAP = 6;
    public static final int PRODUCTIVE_DELTA = 1;
    public static final double OPENING_BALANCER_CHANCE = 0.72;

    private DrawBrain() {}

    /**
     * Serve a card from [stock] to [hand], examining the top {@link
     * #LOOK_AHEAD_WINDOW} entries. Stock uses last-element-as-top ordering
     * (mirrors the ArrayDeque semantics in {@code GameSession}). Removes and
     * returns the chosen card.
     */
    public static Card drawFromStock(List<Card> stock, Hand hand, int dryStreak) {
        return drawFromStock(stock, hand, dryStreak, AdaptiveDrawParams.DEFAULTS);
    }

    /** Phase 4 overload: per-user tuning via [params]. */
    public static Card drawFromStock(List<Card> stock, Hand hand, int dryStreak,
                                     AdaptiveDrawParams params) {
        if (stock.isEmpty()) throw new IllegalStateException("empty stock");
        if (stock.size() == 1) return stock.remove(stock.size() - 1);
        int windowSize = Math.min(params.lookAheadWindow(), stock.size());
        int start = stock.size() - windowSize;
        int bestIdx = stock.size() - 1;
        int bestScore = score(stock.get(bestIdx), hand, dryStreak, params);
        for (int i = start; i < stock.size() - 1; i++) {
            int s = score(stock.get(i), hand, dryStreak, params);
            if (s > bestScore) {
                bestScore = s;
                bestIdx = i;
            }
        }
        return stock.remove(bestIdx);
    }

    /** True if drawing [drawn] into [handBefore] would reduce the optimised count. */
    public static boolean wasProductive(Card drawn, Hand handBefore) {
        int before = ScoringEngine.count(handBefore.cards());
        List<Card> withDraw = new ArrayList<>(handBefore.cards());
        withDraw.add(drawn);
        int bestAfter = Integer.MAX_VALUE;
        for (int i = 0; i < withDraw.size(); i++) {
            List<Card> rest = new ArrayList<>(withDraw);
            rest.remove(i);
            bestAfter = Math.min(bestAfter, ScoringEngine.count(rest));
        }
        return before - bestAfter >= PRODUCTIVE_DELTA;
    }

    /**
     * V2.2 §38–39: if [hand] has no pair, with {@link
     * #OPENING_BALANCER_CHANCE} search [stock] for a matching-rank card and
     * swap it with the hand's highest-value card. Never guarantees a pair;
     * preserves card conservation; only mutates on hit.
     */
    public static void balanceOpeningHand(Hand hand, List<Card> stock, Random rng) {
        balanceOpeningHand(hand, stock, rng, AdaptiveDrawParams.DEFAULTS);
    }

    /** Phase 4 overload: per-user tuning via [params]. */
    public static void balanceOpeningHand(Hand hand, List<Card> stock, Random rng,
                                          AdaptiveDrawParams params) {
        if (hand.size() == 0 || stock.isEmpty()) return;
        if (hasAnyPair(hand)) return;
        if (rng.nextDouble() >= params.openingBalancerChance()) return;

        Set<Rank> ranksHeld = new HashSet<>();
        for (Card c : hand.cards()) ranksHeld.add(c.rank());
        int stockIdx = -1;
        for (int i = stock.size() - 1; i >= 0; i--) {
            Card s = stock.get(i);
            if (!s.isSpecial() && ranksHeld.contains(s.rank())) { stockIdx = i; break; }
        }
        if (stockIdx < 0) return;

        Card victim = null;
        for (Card c : hand.cards()) {
            if (c.isSpecial()) continue;
            if (victim == null || c.value() > victim.value()) victim = c;
        }
        if (victim == null || victim.value() < stock.get(stockIdx).value()) return;

        Card promoted = stock.get(stockIdx);
        hand.remove(victim);
        hand.add(promoted);
        stock.set(stockIdx, victim);
    }

    // ---------- utility scoring (mirrors Dart weights) --------------------

    static int score(Card c, Hand hand, int dryStreak) {
        return score(c, hand, dryStreak, AdaptiveDrawParams.DEFAULTS);
    }

    static int score(Card c, Hand hand, int dryStreak, AdaptiveDrawParams params) {
        Map<Rank, Integer> rankCount = new HashMap<>();
        boolean hasSpecial = false;
        for (Card h : hand.cards()) {
            if (h.isSpecial()) { hasSpecial = true; continue; }
            rankCount.merge(h.rank(), 1, Integer::sum);
        }
        int utility = 0;
        boolean useful = false;

        if (c.isSpecial()) {
            boolean hasExactPair = rankCount.values().stream().anyMatch(n -> n == 2);
            utility += hasExactPair ? (35 + params.specialUtilityBoost()) : -10;
            useful = hasExactPair;
        } else {
            int n = rankCount.getOrDefault(c.rank(), 0);
            if (n >= 2) {
                utility += 40; useful = true;
            } else if (n == 1) {
                utility += 25; useful = true;
                if (hasSpecial) utility += 30 + params.specialUtilityBoost();
            } else if (c.value() <= 3) {
                utility += 6; useful = true;
            } else if (c.value() <= 6) {
                utility += 2;
            } else {
                utility -= 6;
            }
        }
        if (dryStreak >= params.dryDrawThreshold() && useful) {
            int level = Math.min(dryStreak, DRY_DRAW_CAP) - params.dryDrawThreshold() + 1;
            utility += (int) Math.round(level * 8 * params.dryPityMultiplier());
        }
        return utility;
    }

    private static boolean hasAnyPair(Hand hand) {
        Map<Rank, Integer> counts = new HashMap<>();
        for (Card c : hand.cards()) {
            if (c.isSpecial()) continue;
            int n = counts.merge(c.rank(), 1, Integer::sum);
            if (n >= 2) return true;
        }
        return false;
    }
}
